src/frequenz/sdk/actor/_data_sourcing/microgrid_api_source.py

```diff
@@ -343,18 +343,39 @@ async def _handle_data_stream(
             stream_senders = self._get_metric_senders(
                 category, self._req_streaming_metrics[comp_id]
             )
-        api_data_receiver = self.comp_data_receivers[comp_id]
+        api_data_receiver: Receiver[Any] = self.comp_data_receivers[comp_id]
+
+        senders_done: asyncio.Event = asyncio.Event()
+        pending_messages = 0
 
         def process_msg(data: Any) -> None:
             tasks = []
             for extractor, senders in stream_senders:
                 for sender in senders:
                     tasks.append(sender.send(Sample(data.timestamp, extractor(data))))
-            asyncio.gather(*tasks)
+
+            def done_callback(_: asyncio.Future) -> None:
+                nonlocal pending_messages
+                pending_messages -= 1
+                if pending_messages == 0:
+                    senders_done.set()
+
+            nonlocal pending_messages
+            pending_messages += 1
+            asyncio.gather(*tasks).add_done_callback(done_callback)
 
         async for data in api_data_receiver:
+            senders_done.clear()
             process_msg(data)
 
+        while pending_messages > 0:
+            await senders_done.wait()
+
+        for _, requests in self._req_streaming_metrics[comp_id].items():
+            for request in requests:
+                # pylint: disable=protected-access
+                await self._registry._close_channel(request.get_channel_name())
```

Comment:
You don't need to use `.items()` if you don't care about the keys. Also, it would be good to make all those close run in parallel:

```suggestion
        await asyncio.gather(*[
            self._registry._close_channel(r.get_channel_name())
            for requests in d for r in requests])
```

Reply:
Hmm if I don't use `items()` it gives me only the keys, at least if I can believe pyright. 

```
   for requests in self._req_streaming_metrics[comp_id]:                        
       for request in requests:
                      ^^^^^^^^ "ComponentMetricId" is not iterable
                               __iter__" method not defined (Pyright reportGeneralTypeIssues)
```

Reply:
Sorry, for the values you need to use `.values()` :)